### PR TITLE
Userstep

### PR DESCRIPTION
**Pull request by renatoloureiro**
I believe the new step is working fine and well written, the only aspect that I would like to discuss is if it's a good idea that the user defined function receives the variable `prob` every time. Also this new step can only have one output, i.e. one variable as output.

### CODE REVIEW

File: +bisos/+iteration/+step/transfer.m [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+methods
+    function step = transfer(prob, output, fun_handle, input, varargin)
+        % New transfer step.
+        
+        assert(isa(fun_handle, "function_handle"), 'Not a function handle.');
+
+        step.fun = fun_handle; 
+        
+        assert(nargin(fun_handle)-1==length(input),...
+            'Number of inputs is incorrect %d\nShould be %d', length(input), nargin(fun_handle)-1);
```
Comment (tcunis) [inline +bisos/+iteration/+step/transfer.m:19]: Please keep the message for the exception on a single line. Also, be a little more specific: Number of which inputs do you mean?

File: +bisos/+iteration/+step/transfer.m [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        step.varin = input;
+
+        assert(length(output)==1, 'Not a single output');
```
Comment (tcunis) [inline +bisos/+iteration/+step/transfer.m:23]: Be more specific, 
```suggestion
        assert(length(output)==1, 'Function handle must return exactly one output.');
```

File: +bisos/+iteration/+step/transfer.m [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    function [sol,info,stop] = run(step,prob,info,sol,symbols,assigns,options)
+        % Run transfer step.
+        
+        a = cell(1, length(step.varin));
+        for i=1:length(step.varin)
+            a{i} = sol.(step.varin{i});
+        end
+
+        % set objective
+        sol.(step.varout{1}) = step.fun(prob, a{:});
```
Comment (tcunis) [inline +bisos/+iteration/+step/transfer.m:42]: Do not forward problem structure to function handle.

**Review by tcunis — CHANGES_REQUESTED** [review]
We should not provide the problem structure to the user-defined function. I don't think it's good style and may also lead to confusion if the user is not aware of such a first parameter to their function. In order to provide variables beyond the decision variables, I suggest a mechanism similar to `outputfcn`, where the user provides a cell array of arguments when registering the step.

**Comment by renatoloureiro** [comment]
With the latest commit https://github.com/tcunis/bisosprob/pull/20/commits/50a138ed1119e6c2b2d9d5c9c794dfd09986994d , all the previous requested changes have been  handled

**Review by tcunis — APPROVED** [review]
Looks good to me. I have just tidied up the demos. Why do we allow the `transfer` step to only change a single variable?

**Comment by renatoloureiro** [comment]
> Looks good to me. I have just tidied up the demos. Why do we allow the `transfer` step to only change a single variable?

Right now with the current step, it only allows for a single output, i.e. change in only one variable. I understand that we could extend this to multiple outputs instead of just one, but I decided that it would be better to just have one output. So that while using it, the output depends on all inputs, simple example:
 p, V <- b, c, a    (multiple outputs)
equivalent to
p <- b
V <-c, a

all outputs don't necessarily depend on all inputs in the multiple outputs case.

**Pull request merged**